src/client/particles.cpp

```diff
@@ -251,29 +259,53 @@ ParticleSpawner::ParticleSpawner(
 	LocalPlayer *player,
 	const ParticleSpawnerParameters &p,
 	u16 attached_id,
-	video::ITexture *texture,
+	ClientParticleTexture* texpool,
+	size_t texcount,
```

Comment:
please use an [`std::vector`](https://en.cppreference.com/w/cpp/container/vector) instead.

Reply:
is that really wise? vectors are for arrays that change size of the course of their lifetimes. `texpool` will never change size under any circumstances (unless we later decide to add a command to alter live particlespawners), but we'll end up paying for the extra space overhead anyway. also, the accessor operator (which will be called *every time a new particle is spawned*, which could be easily be thousands of times per second even on a very small server, especially iwth weather mods in use) has significant extra overhead beyond the simple add+deref that is currently used -- apart from the overhead of a function call, an extra level of indirection is necessary because `std::vector` has to account for the possibility that its contents will be moved somewhere else in memory between calls, something that will *never happen* to a texpool. and all this overhead will even affect particlespawners that only have a single texture.

Reply:
> vectors are for arrays that change size of the course of their lifetimes. texpool will never change size under any circumstances 

True, `std::array` would be a fixed size type but it has to be fixed at compile time so that won't help here.

> the accessor operator has significant extra overhead beyond the simple add+deref that is currently used

`operator[]` does no bounds-checking, you might be thinking of [`at`](https://en.cppreference.com/w/cpp/container/vector/at). Other than that it should be no different than a C-style indexing.

> apart from the overhead of a function call

Templates are inlined

> an extra level of indirection is necessary because std::vector has to account for the possibility that its contents will be moved somewhere else in memory between calls

I don't know what this is referring to and this is my first time hearing it. A vector internally encapsulates a pointer to a continuous memory region containing its contents, I don't see the difference here: it could be swapped out just like a pointer kept manually.

Reply:
update: okay, i ran some tests. looks like `std::vector` (at least on g++) is nowhere near as bad as it was when i was 17, but there's still a nearly 2x slowdown compared to normal array accesses:

```c++
$ cat vectest.cc
#include <vector>
#include <iostream>

int main(int argc, char** argv) {
	std::vector<int> vec = {1,2,3,4,5,6,7,8,9,10,11};
	int* arptr = new int[11]{1,2,3,4,5,6,7,8,9,10,11};
	size_t q = 0;
	for (size_t i = 0; i < 1ULL<<32; ++i)
#		ifdef test_vector
			q += vec[argc];
#		else
			q += arptr[argc];
#		endif

	return q;
}

$ g++ vectest.cc -ovectest -Dtest_vector
$ g++ vectest.cc -optrtest
$ time ./ptrtest 1

________________________________________________________
Executed in    7.04 secs   fish           external 
   usr time    7.03 secs  798.00 micros    7.03 secs 
   sys time    0.00 secs  214.00 micros    0.00 secs 

$ time ./vectest 1

________________________________________________________
Executed in   12.31 secs   fish           external 
   usr time   12.31 secs    0.00 micros   12.31 secs 
   sys time    0.00 secs  1079.00 micros    0.00 secs 
```

(i also tried compiling the tests with -O4 but g++ is clearly outsmarting me in that case, because ever absurd numbers like `1LL<<63` finish immediately under both implementations; i looked at them with a disassembler and there's no loop at all in those versions)

Reply:
Hmm that's unexpected to me.
Testing a different example that can't be optimized away (a sum) I found a case where assembly for `std::vector` and `int*` are exactly identical: https://godbolt.org/z/c8bazY67o

> (i also tried compiling the tests with -O4 but g++ is clearly outsmarting me in that case, because ever absurd numbers like 1LL<<63 finish immediately under both implementations; i looked at them with a disassembler and there's no loop at all in those versions)

I tested that and even with `-O2` both gcc and clang remove(?) the loop here.

　

In any case we use vectors and other C++ containers a lot since they're less error prone and more convenient to work with than pointers, so if the overhead (which exists depending on how you use them) breaks performance on someone's machine I bet they weren't able to run Minetest anyway.

Reply:
> Hmm that's unexpected to me.

take a look at `echo "#include <vector>" | g++ -xc++ -E -P` and you'll probably get a general sense at least of why this might be haha (if you can dig through all the abstraction and template magic; i got lost somewhere around `_Vector_impl_data`)

> Testing a different example that can't be optimized away (a sum) I found a case where assembly for std::vector and int* are exactly identical: https://godbolt.org/z/c8bazY67o

it's interesting but my test shows that the problem *can* occur, which is what's important; it's unwise to rely too heavily on compiler optimizations as an excuse to not just do the right thing in the first place. minetest is much much muuuch more complex than these little examples, and we wouldn't be able to know to what degree a given compiler could optimize away the vector baggage without digging into the assembly of each one (which, ouch) or how those optimizations might change in future versions of the compiler (they're not part of the c++11 standard, after all). and here's the thing: even in the most optimistic case, where the vector access is boiled down to exactly the same code as the array access, all it takes is one unsuspecting programmer a few years down the road making a call to some particular vector function that arrays don't have (say `push_back()`) to potentially cause an explosion of complexity by invalidating optimisations we were relying on; can you imagine trying to track down something like that?

so i'm really sorry but i just don't think it's acceptable to further worsen particle performance (even if only under certain compilers/stdlibs!) which, as hecktest has pointed out, is already pretty bad, just for a mildly convenient new feature (it would be different if these were necessary for something very important of course), especially as you seemingly (apologies if i've misunderstood or missed something) have no specific reason for wanting to use vectors in a performance-critical part of the code and vector features are of no use here. it's a bit like using chromium instead of `cat` to read to a text file (huge exaggeration ofc, but you get my drift). "we can probably get away with it" is a far cry from "we need to," and having succumbed to that mindset on many occasions, i have learned the hard way that these things add up -- and none of those projects were anywhere near the size and complexity of minetest. it's easier to speed up code when there's ten obvious inefficencies, much much harder when there's a thousand tiny little inefficiencies that would all be fine on their own but together bring a system to its knees, especially because they're so damn hard to find once they've been written.

(for example: i noticed last night that `ParticleManager` does not preallocate space (i.e. with `reserve`) when a new particlemanager is added! this completely escaped my notice until now, even tho i've been tits-deep in the particle engine code off and on for two weeks. i've started on a fix but it needs a tiny bit of refinement to handle edge cases before it can be merged)

bottom line: if you have a *concrete* objection to pointers (as you yourself have indicated, vectors do not perform bounds checking, so they are no less error-prone than pointers in cases where the arrays do not change over their lifetime), i would very much like to work with you to figure out an optimal solution! for instance, if your concern is that raw malloc'd arrays don't cooperate with the RAII idiom that this project makes extensive use of (which is a very reasonable objection! i did have to add a destructor that wasn't there before to accomodate this array), a simple `flexArray<T>` template that behaves like `std::arrray<T,N>` but uses heap storage and allows arbitrary sizes would be a perfect compromise; it would address the actual problem without impacting performance in the slightest.

if that's not good enough for you, or this is just about discomfort with manual array use rather than a concrete problem arrays introduce, then tbh i believe the right thing to do is to rip out texpool support and just return to one texture per emitter. if use of vectors had been the only way to implement this functionality, i would never have added it in the first place. rubenwardy's initial reaction to #11505 was wise and very much warranted -- i don't think i should be cramming in new features that hit performance-critical areas, even slightly, without extremely good reason; i already killed particle texture blending over this and this isn't nearly as useful. so i'm not interested in being responsible for further slowing the engine down over it, not even mildly. especially since adding actually important features has already required non-negligible structure size increases (affecting both the network protocol and local memory on client and server) and further complications of code that has to run many times each second on the client. a visual equivalent to texpool support could be implemented purely in a couple lines of lua if we wanted to; there's no real reason it *has* to be in the engine… and perhaps most importantly, i'd rather focus my attention on the features that motivated me to make this PR in the first place haha :)

Reply:
Well it doesn't look like we're coming to an agreement here. Essentially it's a matter of whether it is premature optimization or not, I believe it is and continue to insist that `std::vector` is the better alternative.

If other devs chime in a decision could be made but that can wait.

> particle performance which, as hecktest has pointed out, is already pretty bad

His remark was regarding the rendering performance of particles, which suffers because vertices are constantly being modified. You didn't add more of that so don't worry.

> if [...], then tbh i believe the right thing to do is to rip out texpool support and just return to one texture per emitter.

I don't think so. Even if texture pools can be replicated by creating multiple particle spawners with differing parameters from Lua it's still an useful feature IMO.

> i'd rather focus my attention on the features that motivated me to make this PR in the first place haha :)

Please do.

Reply:
update: hopefully this will be an acceptable compromise? i've removed the raw array and switched to an `std::unique_ptr`, this provides the benefits of using a vector without most of the unnecessary baggage or capabilities that aren't needed

Reply:
> If other devs chime in a decision could be made but that can wait.

alright. i'll leave the code in place pending; hopefully we can come to an agreement on a more appropriate solution than `std::vector`. failing that, i ask that this PR not be merged until i've had time to remove the texpool code; i hope you can understand why i'm not interested in bearing the `git praise` for creating what i consider to be needless technical debt.